File: server/controller/controller.ts
import { Request, Response } from "express";
import { Session, ShopList } from "./models";

const maxSessions = 10_000; // simple protection

export class Controller {
  sessions: Map<string, Session> = new Map();

  /**
     CreateSession initie une nouvelle session de courses,
     avec la liste donnée.
     */
  CreateSession = (req: Request, res: Response) => {
    const list = req.body as ShopList;
    const id = this.createSession(list);
    res.json({ sessionID: id });
  };

  // renvoie l'id créé
  private createSession(list: ShopList): string {
    if (this.sessions.size >= maxSessions) {
      throw "internal error: maximum number of session reached";
    }
    let id = randString();
    while (this.sessions.has(id)) {
      id = randString();
    }
    const s: Session = { id: id, list: list };

    this.sessions.set(id, s);

    console.log(`Creating session ${id} with ${list.length} ingredients`);

    return id;
  }

  GetSession = (req: Request, res: Response) => {
    const id = (req.query["sessionID"] || "") as string;
    const session = this.sessions.get(id);
    if (session === undefined) {
      throw `La session <${id}> est invalide ou terminée.`;
    }
    res.json(session.list);
  };

  UpdateSession = (req: Request, res: Response) => {
    const args = req.body as update;

    const id = (req.query["sessionID"] || "") as string;
    const l = this.updateSession(args, id);

    res.json(l.list);
  };

  private updateSession(args: update, sessionID: string): Session {
    const l = this.sessions.get(sessionID);
    if (l === undefined) {
      throw `La session <${sessionID}> est invalide ou terminée.`;
    }

    l.list.forEach((v, i) => {
      if (v.id == args.id) {
        l.list[i].checked = args.checked;
      }
    });

    return l;
  }
}

type update = {
  checked: boolean;
  id: number;
};

function randString(): string {
  const L = 16;
  let result = "";
  const characters =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  const charactersLength = characters.length;
  let counter = 0;
  while (counter < L) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
    counter += 1;
  }
  return result;
}

File: server/src/controllers/shop-session/controller.go
package shopsession

import (
	"errors"
	"fmt"
	"sync"

	"github.com/benoitkugler/atable/utils"
	"github.com/labstack/echo/v4"
)

const maxSessions = 10_000 // simple protection

type Controller struct {
	sessions map[string]Session
	lock     sync.Mutex
}

func NewController() *Controller {
	return &Controller{sessions: map[string]Session{}}
}

// CreateSession initie une nouvelle session de courses,
// avec la liste donnée.
func (ct *Controller) CreateSession(c echo.Context) error {
	var args ShopList
	if err := c.Bind(&args); err != nil {
		return err
	}
	out, err := ct.createSession(args)
	if err != nil {
		return err
	}
	return c.JSON(200, out)
}

// renvoie l'id créé
func (ct *Controller) createSession(list ShopList) (CreateSessionOut, error) {
	ct.lock.Lock()
	defer ct.lock.Unlock()

	if len(ct.sessions) >= maxSessions {
		return CreateSessionOut{}, errors.New("internal error: maximum number of session reached")
	}
	id := utils.RandomString(false, 10)
	for _, has := ct.sessions[id]; has; _, has = ct.sessions[id] {
		id = utils.RandomString(false, 10)
	}
	s := Session{Id: id, List: list}

	ct.sessions[id] = s

	fmt.Printf("Creating session %s with %d ingredients\n", id, len(list))

	return CreateSessionOut{SessionID: id}, nil
}

func (ct *Controller) GetSession(c echo.Context) error {
	id := c.QueryParam("sessionID")
	session, ok := ct.sessions[id]
	if !ok {
		return fmt.Errorf("La session <%s> est invalide ou terminée.", id)
	}
	return c.JSON(200, session)
}

func (ct *Controller) UpdateSession(c echo.Context) error {
	var args UpdateSessionIn
	if err := c.Bind(&args); err != nil {
		return err
	}

	id := c.QueryParam("sessionID")
	out, err := ct.updateSession(args, id)
	if err != nil {
		return err
	}

	return c.JSON(200, out)
}

func (ct *Controller) updateSession(args UpdateSessionIn, sessionID string) (Session, error) {
	ct.lock.Lock()
	defer ct.lock.Unlock()

	session, ok := ct.sessions[sessionID]
	if !ok {
		return session, fmt.Errorf("La session <%s> est invalide ou terminée.", sessionID)
	}

	for i, v := range session.List {
		if v.Ingredient.Id == args.Id {
			session.List[i].Checked = args.Checked
		}
	}

	return session, nil
}
